src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResultBuilder.cs

```diff
@@ -122,31 +158,65 @@ internal void ApplySuppressionsAndStoreAnalysisResult(AnalysisScope analysisScop
                     }
                 }
             }
+
+            static SyntaxTree? getSourceTree(Diagnostic diagnostic)
+                => diagnostic.Location.SourceTree;
+
+            AdditionalText? getAdditionalTextKey(Diagnostic diagnostic)
+            {
+                // Fetch the first additional file that matches diagnostic location.
+                if (diagnostic.Location is ExternalFileLocation externalFileLocation)
+                {
+                    var path = externalFileLocation.FilePath;
+                    if (_pathToAdditionalTextMap.TryGetValue(externalFileLocation.FilePath, out var additionalTexts))
+                    {
+                        if (additionalTexts.Count == 1)
+                        {
+                            return additionalTexts[0];
```

Comment:
>return additionalTexts[0]; [](start = 28, length = 26)

Do we need to check `analysisScope.NonSourceFiles.Contains(...)` here? If so, please remove this special case and handle in `foreach` below.